src/backend/distributed/transaction/distributed_transaction_management.c

```diff
@@ -0,0 +1,433 @@
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * get_distributed_transaction_id returns a tuple with (databaseId, initiatorNodeIdentifier,
+ * transactionId, timestamp) that exists in the shared memory associated with this backend.
+ */
+Datum
+get_distributed_transaction_id(PG_FUNCTION_ARGS)
+{
+	Datum distributedTransactionId = 0;
+
+	Oid databaseId = InvalidOid;
+	uint64 initiatorNodeIdentifier = 0;
+	uint64 transactionId = 0;
+	TimestampTz timestamp = 0;
+
+	CheckCitusVersion(ERROR);
+
+	if (!DistributedTransactionBackend)
+	{
+		InitializeDistributedTransactionManagementBackend();
+	}
+
+	/* read the shmem while mutex is held */
+	SpinLockAcquire(&BackendDistributedTransactionMutex);
+
+	databaseId = DistributedTransactionBackend->databaseId;
+	initiatorNodeIdentifier =
+		DistributedTransactionBackend->transactionId.initiatorNodeIdentifier;
+	transactionId = DistributedTransactionBackend->transactionId.transactionId;
+	timestamp = DistributedTransactionBackend->transactionId.timestamp;
+
+	SpinLockRelease(&BackendDistributedTransactionMutex);
+
+	distributedTransactionId =
+		GenerateDistributedTransactionIdTuple(databaseId, initiatorNodeIdentifier,
+											  transactionId, timestamp);
+
+	PG_RETURN_DATUM(distributedTransactionId);
+}
+
+
+/*
+ * GetAllActiveDistributedTransactions returns all the active distributed
+ * transactions on the shared memory.
+ *
+ * Note that the function could not gurantee a global consistent result, the function
+ * only locks each active backend while copying it. Currently the only caller of this
+ * function is in the src/test folder. The consistency gurantee is sufficient for
+ * testing.
+ */
+List *
+GetAllActiveDistributedTransactions(void)
+{
+	int sessionIndex = 0;
+	List *activeTransactionList = NIL;
+
+	for (sessionIndex = 0; sessionIndex < MaxBackends; ++sessionIndex)
+	{
+		DistributedTransactionBackendData *currentBackend =
+			&distributedTransactionShmemData->sessions[sessionIndex];
+
+		if (currentBackend && currentBackend->transactionId.transactionId != 0)
+		{
+			DistributedTransactionBackendData *activeBackend =
+				(DistributedTransactionBackendData *) palloc(
+					sizeof(DistributedTransactionBackendData));
+
+			SpinLockAcquire(&BackendDistributedTransactionMutex);
+
+			activeBackend->databaseId = currentBackend->databaseId;
+			activeBackend->transactionId.initiatorNodeIdentifier =
+				currentBackend->transactionId.initiatorNodeIdentifier;
+			activeBackend->transactionId.transactionId =
+				currentBackend->transactionId.transactionId;
+			activeBackend->transactionId.timestamp =
+				currentBackend->transactionId.timestamp;
+
+			SpinLockRelease(&BackendDistributedTransactionMutex);
+
+			activeTransactionList = lappend(activeTransactionList, currentBackend);
+		}
+	}
+
+	return activeTransactionList;
+}
+
+
+/*
+ * GenerateDistributedTransactionIdTuple returns a datum of psudo-generated heaptuple
+ * for the distributed transaction id provided by the parameters.
+ */
```

Comment:
generated: All heap tuples are "generated"
pseudo: It can be returned to the caller and the user, so it's not really pseudo or some such. It's not faking being a tuple.